lib/puppet/util/posix.rb

```diff
@@ -1,6 +1,39 @@
 # Utility methods for interacting with POSIX objects; mostly user and group
 module Puppet::Util::POSIX
 
+  module GrpLibC
+    def self.user_groups(user)
+      result = []
+      ngroups = 16
+
+      FFI::MemoryPointer.new(:int) do |ngroups_ptr|
+        FFI::MemoryPointer.new(:uint, ngroups) do |groups_ptr|
+          ngroups_ptr.write_int(ngroups)
+          user_gid = Puppet::Etc.getpwnam(user).gid
+
+          data = Puppet::FFI::Posix::Functions::getgrouplist(user, user_gid, groups_ptr, ngroups_ptr)
+
+          if data < 0
+            groups_ptr.free if groups_ptr
+            FFI::MemoryPointer.new(:uint, ngroups_ptr.read_int) do |new_groups_ptr|
+              Puppet::FFI::Posix::Functions::getgrouplist(user, user_gid, new_groups_ptr, ngroups_ptr)
+              groups_ptr = new_groups_ptr
+            end
+          end
+
+          groups_gids = groups_ptr.get_array_of_uint(0, ngroups_ptr.read_int)
+
+          groups_gids.each do |group_gid| 
+            group_info = Puppet::Etc.getgrgid(group_gid)
+            result.push(group_info.name) if group_info.mem.include?(user)
```

Comment:
On OS X I had the surprise of having multiple groups of the same name, maybe this is also somehow possible with AD groups, in which case I'd change this line to:

```ruby
result |= [group_info.name] if group_info.mem.include?(user)
```

This way we only append the group if it doesn't exist in `result`. I'm fairly sure this edge case is wrong and almost impossible to see in practice, but... you never know.

Reply:
Added your variant as extra protection and future proofing but the ending of `groups_of` method where the list of groups, in both implementations, are filtered for duplicates with `*.uniq` should've covered this case already.